### DISCUSSION TITLE
Matching errors without checking NSError's userInfo dictionary

### DISCUSSION THREAD

**Discussion by csiszarattila** [Quick/Nimble#904] [Ideas]
Hi,

I try to write a simple match for an UrlError:

```
  expect(error).to(
     matchError(NSError(domain: NSURLErrorDomain, code: URLError.timedOut.rawValue, userInfo: nil)
  )
```

But it fails, even the error is the same,  because matchError seems to also compare the userInfo:

expected to match error <Error Domain=NSURLErrorDomain Code=-1001 "(null)">, got <Error Domain=NSURLErrorDomain Code=-1001 "The request timed out." UserInfo={NSUnderlyingError=0x600001431260 {Error Domain=kCFErrorDomainCFNetwork Code=-1001 "(null)" UserInfo={_kCFStreamErrorCodeKey=-2102, _kCFStreamErrorDomainKey=4}}, NSErrorFailingURLStringKey=https://example.org/example/end/point, NSErrorFailingURLKey=https://example.org/example/end/point, _kCFStreamErrorDomainKey=4, _kCFStreamErrorCodeKey=-2102, NSLocalizedDescription=The request timed out.}>

Is it intentional? Is there a way to overcome this?

Thanks.

**Comment by ikesyo** [comment]
I'm sorry for the late response.

I think that's because `NSError` equality check is implemented as such and the overload is used: https://github.com/Quick/Nimble/blob/025d0e25fc043534ae24e645e6729cd15851e28d/Sources/Nimble/Matchers/MatchError.swift#L34-L53

So I'd say this is an expected behavior.

But it may be good to provide another overload which takes an `NSError` and a matching strategy such as:

```swift
enum NSErrorMatchingStrategy {
    case `default` // Rely on `NSError`'s `Equatable` implementation
    case ignoringUserInfo // Just compare `domain` and `code` and ignore `userInfo`
}

func matchError(_ error: NSError, strategy: NSErrorMatchingStrategy = .default) -> Predicate<NSError>
```
